Multiple roscore

joshuamorton opened the pull request:
This partially addresses #177, although it doesn't close it and isn't complete anyway.

(Actually it may completely address it).

Additionally, we want to fiddle with the mock_node and mock topic files to make them work on repeated use, which they currently don't. 

That would mean changing from tearing the topic down to using a new subscriber, but the same node, for each test.

joshuamorton committed:
This should break all tests.

joshuamorton committed:
it works I think, hacky as it may be.

irapha commented on buzzmobile/tests/test_utils/rostest_utils.py:
```diff
@@ -1,38 +1,74 @@
+        def new_setup(self):
+            """Wrapper around the user-defined setUp method that runs roscore.
+            """
+            self.port = rand_port()
```
Doesn't this have the potential for pick the same port number for two parallel tests (thus causing issues)?

joshuamorton replied:
yes, hence the todo to retry until success, which should just be (in pseudocode):

    port = randport
    while port is in use:
        port = randport

irapha replied:
I definitely overlooked the TODO. My bad

irapha commented on buzzmobile/tests/test_utils/rostest_utils.py:
```diff
@@ -1,38 +1,74 @@
+        # the real ones don't exist yet.
+
+        def fake_settear(self):
+            _ = self
```
Can't we just `pass`?

irapha replied:
Or use a `lambda self: pass` in lines 30 and 34?

joshuamorton replied:
`lambda self: pass` is a syntax error (pass inside a lambda), and 

    def func(self):
        pass

gets yelled at by style for an unused variable (self).

irapha replied:
How about
```python
def noop(_):
    pass
```

irapha replied:
Or `lambda _: True`

joshuamorton replied:
Either of those would work I think, `noop` seems like the better solution in this case, the lambda feels sneaky.

joshuamorton replied:
done locally.

irapha commented on buzzmobile/tests/test_utils/rostest_utils.py:
```diff
@@ -1,38 +1,74 @@
+        dct['setUp'] = new_setup
+        dct['tearDown'] = new_teardown
+        dct['setUp'].__name__ = 'setUp'
+        dct['tearDown'].__name__ = 'tearDown'
```
Isn't this `__name__` setting easier with `functools.wraps`?

joshuamorton replied:
unfortunately that doesn't work if we're injecting our own fake setup or teardown function (ie. the mock_settear one).

irapha replied:
Alright

irapha commented on buzzmobile/tests/unit/process/test_bearing.py:
```diff
@@ -16,8 +16,9 @@ def test_distance(self):
-class TestBearingNode(unittest.TestCase):
+class TestBearingNode(RosTest):
+    def setUp(self):
+        pass
```
Since you have a fake setup, do you need this empty one here?

joshuamorton replied:
no, I can (and will) remove it. I was testing some metaclass errors I was encountering related to the class dict.

joshuamorton replied:
done locally.

irapha commented on buzzmobile/tests/test_utils/launch_tools.py:
```diff
@@ -70,22 +77,23 @@ def launcher(func):
-            node = roslaunch.core.Node(package, name, namespace=namespace)
+            env = {'ROS_MASTER_URI': self.rosmaster_uri}
+            node = roslaunch.core.Node(package, name, namespace=namespace,
+                    env_args=env.iteritems())
```
This `env_args`; could it be the key to fixing #157 ?

joshuamorton replied:
Unfortunately, I don't *think* so. env args sets environmental variables (equivalent to `os.environ[x]= y` in python, or `export VAR=VALUE` in bash). I'll expand on this in a sec, but it doesn't do anything with command line args.

irapha replied:
Oh ok. I also misread `env_args` as `cmd_args` this morning when I was reviewing

irapha commented on buzzmobile/tests/test_utils/launch_tools.py:
```diff
@@ -44,13 +51,13 @@ def launcher(func):
             """
-            launch = ROSLauncher(full_name)
+            os.environ['ROS_MASTER_URI'] = self.rosmaster_uri
+            launch = ROSLauncher(full_name, port=self.port)
```
Why is the creation of a launcher different between here and `launch_node`? Couldn't each of these decorators theoretically be called independently or the other?

irapha replied:
Shouldn't we, for instance, also be checking whether `_LAUNCHER[self.port]` already exists?

irapha replied:
This is esp weird since `ROSLauncher.__init__` doesn't seem to return anything

joshuamorton replied:
`__init__` never returns anything ;)

the broader question, we want to allow all of the following:

    @with_launch_file(*args)
    @mock_node(*node_args)
    @mock_node(*other_args)
    @mock_node(*more_of_these)
    def test func():
        ...

as well as

    @mock_node(*some_args)
    @mock_node(*diff_args)
    def test_function():
        ...

Note that in the second case with_launch is never called. This might be the case if there was no initialization, but in both cases, we need to initialize a "master" launcher for the test through which all nodes are launched. With `with_launch_file`, we can be sure that it will be the first (and last) thing to be executed wrapping the function, so it can *always* take over the port.

Mock_node on the other hand might sometimes need to do it, but only if there is no existing launcher, in which case it needs to know that it is in charge and clean up after itself (shutdown the launcher).

I can add that in as a comment if you want.

irapha replied:
> `__init__` never returns anything ;)

I think I was tired when I wrote that question

irapha replied:
If I understand correctly, this means that while both of the examples you wrote are allowed, the following would *not* be:

```python
@mock_node(*node_args)
@mock_node(*other_args)
@mock_node(*more_of_these)
@with_launch_file(*args)
def test func():
    ...
```

or 

```python
@mock_node(*node_args)
@mock_node(*other_args)
@with_launch_file(*args)
@mock_node(*more_of_these)
def test func():
    ...
```

Because `with_launch_file` assumes it is the outter wrap, thus it can assume no launcher exists right?

And yeah can you make a comment for that?

joshuamorton replied:
Correct. Essentially, the rule is that `with_launch_file` needs to be first, otherwise you get undefined behavior (it might also be good to stick some error handling in there tbh).

joshuamorton replied:
done locally.

irapha reviewed (changes requested):
There's also style changes to be made

In the meantime, I'll test this locally :)

irapha commented:
Another question I had was whether we needed to go back to doing `pytest test/unit` instead of the boxed version.. If so, we should remove the `pytest-xdist` dep

irapha commented:
Local tests are very flaky. Using both `pytest tests/unit` and `pytest -n1 tests/unit --boxed`.

Also, some clean up doesn't seem to be happening. One of my tests failed and it output the message: `ROS_MASTER_URI was: http://ubuntu:12053/`. So I try to run `roscore -p 12053` which gives me:

```
roscore cannot run as another roscore/master is already running. 
Please kill other roscore/master processes before relaunching.
The ROS_MASTER_URI is http://ubuntu:12053/
The traceback for the exception was written to the log file
```

Also, I did: `killall -9 rosmaster` and then `roscore -p 12053` and got a warning: `WARNING: ROS_MASTER_URI port [11311] does not match this roscore [12053]`. But I believe this is being handled correctly in test (by looking at the code)

joshuamorton commented:
>Another question I had was whether we needed to go back to doing `pytest test/unit` instead of the boxed version.. If so, we should remove the `pytest-xdist` dep

We shouldn't *need* to. I'm not sure what effect it will have though. Staying shouldn't hurt.

>Also, I did: `killall -9 rosmaster` and then `roscore -p 12053` and got a warning: `WARNING: ROS_MASTER_URI port [11311] does not match this roscore [12053]`. But I believe this is being handled correctly in test (by looking at the code)

Two things here:
 - if we want multiprocess to work, we can't use `killall -9 rosmaster` anywhere, because it will sigterm all running ros processes and our concurrent tests will all die. Please feel free to stick it in as a stopgap/see if it helps, but that's the downside. (or at least that's the case I think)
 - to run ros with a non-default rosmaster, you need to do two things:
    1. pass the `-p <PORT_NUMBER>` param to roscore
    2. have the environment variable `$ROS_MASTER_URI` set to `http://<hostname>:<PORT_NUMBER>`, where hostname is almost assuredly `localhost`, and `PORT_NUMBER` matches the -p flag
   
   The `env_vars` thing is doing the second one every time I call roscore anywhere. Unfortunately, I can't pass the variable to all of ros's tools (the command line tools like `roslaunch`, `rosgraph`, and `rostopic` don't expose good apis, instead a lot of the public facing functions always and only get the URI from the environment). That's why I have `launch_tools.py:54` in there. (I'd prefer to factor that out into a context manager that sets the nondefault port, runs some function in the context, and then resets the default port, it feels....safer that way, and we'd need to explicitly state where all nondefault port usage was or get errors.

irapha commented:
>>back to doing `pytest test/unit`

>We shouldn't need to.

Well if we want parallelized tests then `-n1 --boxed` prevents that from happening, no?

> if we want multiprocess to work, we can't use `killall -9 rosmaster`

I didn't imply we should use it in test. I was just noting that, after the test was done (and failed), the rosmasters weren't killed. I had to manually kill them with `killall`

irapha committed:
remove I0011 and I0020 (local disable and local supress)

irapha committed:
Merge branch 'master' into multiple-roscore

irapha committed:
add dep instructions to readme

joshuamorton committed:
Merge branch 'master' of github.com:gtagency/buzzmobile into multiple-roscore

joshuamorton committed:
Attempting to fix flakyness on VMs

joshuamorton committed:
Attempting to fix flakyness on VMs.

VM hostnames are not `localhost`, as they are on most systems, this
changes the hostname from being hardcoded as localhost to be actually
defined by the local system.

joshuamorton committed:
Merge branch 'multiple-roscore' of github.com:gtagency/buzzmobile into multiple-roscore

joshuamorton committed:
Merge branch 'multiple-roscore' of github.com:gtagency/buzzmobile into multiple-roscore

joshuamorton committed:
Rename noop setup to 'noop'.

joshuamorton committed:
Removed debugging setUp.

joshuamorton committed:
Added docs for how launch_tools work.

joshuamorton committed:
Style fixes.

joshuamorton committed:
Testing some changes.

joshuamorton committed:
A number of minor changes.

All of these should improve reliability, but none of them seem to do the
trick and actually fix the unreliability of the tests.

joshuamorton committed:
commiting this so its in history.

joshuamorton commented:
I've made a mistake, but a fruitful one:

The only issue that remains is that in `rospy.init_node()`, when called after the first time, the [two calls to `Service` at the very end break](http://docs.ros.org/jade/api/rospy/html/rospy.client-pysrc.html#init_node). This is indicative of a larger issue though, that there really is no way to run rospy.init_node more than once in a safe manner. Essentially, if the main thread every imports rospy anywhere, we're doomed. 

(what's happening by the way is that rospy is saving the get_loggers service url from the first run and using it on subsequent runs, which is problematic because the service doesn't exist because roscore has been killed. to check this, you can delete all contents of `~/.ros/log`, run the tests once, and then grep for "TCP/IP", and check to see that the port number on that line is used in both tests. It should be the only port where that's the case.)

Anyway, rospy can't be used in the main thread. That means that mock_pub, mock_node, etc. need to be reworked so that they have two parts: the python objects that you interface with, and child shell friends that actually import rospy. My plan at this point is a gross amalgam of the pickle library, subprocess, and some impressive dynamism. 

The upside is that this will get us about 95% of the way to #177 being all the way done, I think the only obstacles will be figuring out how to allow this new version of mock_pub and check_topic to access the port information from the outer scope.

joshuamorton committed:
Tests should no longer be flaky.

Thi finally makes everything work!

Creating listener and contoller allows us to run the nodes in background
processes, which doesn't run into rospy issues thankfully. There are a
few changes that need to be made w.r.t. thread safety, currently I do a
hack where I set the environmental variable, which needs to be done, but
that is also used later, and I'd much rather explicitly pass the port to
the context managers than use the shell to get it there if we end up
with multiple processes, although IDK if it will matter.

joshuamorton committed:
Clean up imports.

joshuamorton committed:
Installed psutil.

joshuamorton committed:
Comply with style.

irapha commented on buzzmobile/tests/unit/process/test_bearing.py:
```diff
@@ -16,20 +16,16 @@ def test_distance(self):
     @launch_node('buzzmobile', 'bearing.py')
     def test_bearing_node(self):
-        with mock_pub('/fix', NavSatFix, queue_size=None) as fix_node:
+        with mock_pub('/fix', NavSatFix, queue_size=0) as fix_node:
```
Was this change accidental?

joshuamorton replied:
No, `None` is technically bad, but its not a big issue, so I just swapped it bc y not.

irapha replied:
In that case can you swap in the other test file?

irapha commented on buzzmobile/tests/unit/process/test_bearing.py:
```diff
@@ -16,20 +16,16 @@ def test_distance(self):
                 fix_node.send(NavSatFix(None, None, 39.029128, -111.838257, None, None, None))
 
                 # check the output from the node
-                await(ct.wait_for_message())
```
Do you not await anymore? How can you guarantee a msg has been received?

irapha replied:
Is the answer "because we are trying to read from stdin and until we are able to do so the `.message` property will block"?

irapha replied:
i guess the real issue is there is no way to timeout. I can believe that tests aren't flaky anymore but if there's something weird going on and a msg was published before a listener started listening, then it'll hang forever.

I'm not saying your code is wrong in this respect. I like this solution (as gross as reading from stdin is), but I think we should still have some way to timeout and display a message ("timeout, tests can be flaky: try running once again and if it still doesnt work, investigate")

joshuamorton replied:
You got it right. Essentially, proc.stdout.read will block until the data is there and flushed, which will happen only after its been printed. This blocks until the process is killed I think.

I should be able to rework it to have a timer or timeout, but it'll be even hackier. *shrug*

joshuamorton replied:
see this:
https://stackoverflow.com/questions/1191374/using-module-subprocess-with-timeout

irapha replied:
Well I think this is higher priority than the other things I've commented on in this PR. Without good timeout checking, ci tests could potentially run forever and we'd never notice

joshuamorton replied:
done

irapha commented on buzzmobile/tests/unit/plan/test_controller.py:
```diff
@@ -1,19 +1,16 @@
-         with mock_pub('/joy', Joy, queue_size=None) as joy_node:
+        with mock_pub('/joy', Joy, queue_size=None) as joy_node:
 
             # TODO When this line is enabled, it causes the bearing test to
```
If things are fixed, can you remove this TODO

joshuamorton replied:
done.

irapha commented on buzzmobile/tests/test_utils/rostest_utils.py:
```diff
@@ -1,55 +1,97 @@
+            env = {k:v for k, v in os.environ.iteritems()}
+            env.update({'ROS_MASTER_URI': self.rosmaster_uri})
+            roscore_initialized = False
+            while not roscore_initialized:
```
this takes care of conflicting port numbers correct?

joshuamorton replied:
It should, yes.

irapha commented on buzzmobile/tests/test_utils/mock_nodes.py:
```diff
@@ -4,114 +4,112 @@
-        self.received = False
+        # get port and rosmaster uri cleanly
+
+        # do this better, somehow!
```
add "TODO(joshuamorton)" and an issue

joshuamorton replied:
Interestingly, as I was thinking about it, getting the port and uri may be unnecessary because its already been set.

Alternaitively, I can pull mock_node and check_topic into the RosTest class, and have them reference the port and rosmaster uri set there. Would that be preferable?

irapha replied:
If it makes things cleaner, I don't see why not do it

Does it mean that we'd have to:

```python
@RosTest.with_launch_file(stuff)
@RosTest.launch_node(stuff)
def test(self):
    RosTest.mock_pub(stuff):
        ...
```

cuz that's pretty gross

joshuamorton replied:
nah

    ...in a class
    @with_launch
    @launch_node
    def test(self):
        self.mock_pub(stuff)

irapha replied:
then sure why not

irapha replied:
I'm guessing this was done?

irapha commented on buzzmobile/tests/test_utils/mock_nodes.py:
```diff
@@ -4,114 +4,112 @@
+    def kill(self):
+        """Kills the subscriber ros node process.
+        """
+        self.proc.kill()
```
does this wait until its killed?

joshuamorton replied:
No, but I've fixed that.

irapha commented on buzzmobile/tests/test_utils/mock_nodes.py:
```diff
@@ -4,114 +4,112 @@
+            msg.serialize(sio)
+            data = self.proc.stdout.read(sio.len)
+            msg.deserialize(data)
+            self._message = msg
```
Without the `await` functionality, how can we make sure the msg was received?

irapha replied:
Also, with all of these nodes reading and writing to stdout/in, can you make controller test send and receive a message, just as a sanity check that everything works?

joshuamorton replied:
self.proc.stdout.read is a blocking call that will only return once we have read sio.len bytes, which in this case happens to be exactly the length of a serialized ros message.

irapha replied:
Sure but can you make controller_test send and receive a message, just as a sanity check that everything works when multiple tests are doing so?

joshuamorton replied:
I wanted to, but when I was trying to send a message earlier, the controller node segfaulted.

irapha replied:
maybe make another test for a different node? I just want to make sure it works

irapha commented on buzzmobile/tests/test_utils/mock_nodes.py:
```diff
@@ -4,114 +4,112 @@
+
+        # do this better, somehow!
+        location = './tests/test_utils/listener.py'
+        self.proc = subprocess.Popen([location,
```
Can you make some comments explaining why we are not using roslaunch's built in stuff for spinning up nodes, and why we're doing this separate process thing?

"because this way we guarantee that init_node will be called cleanly and this makes sure that...."

irapha replied:
Generally though, because this is such a weird system, can you make a `buzzmobile/tests/test_utils/readme.md` explaining how it works?

"design doc pls mr intern"

irapha replied:
I got it started for you :)

irapha commented on buzzmobile/tests/test_utils/listener.py:
```diff
@@ -0,0 +1,28 @@
+        """A default callback for our subscriber.
+        """
+        message.serialize(sys.stdout)
+        rospy.signal_shutdown('message sent')
```
im guessing this line is what makes sure things work right?

Is there a way to make a listener be able to receive multiple messages? Not super high priority but it would be nice

joshuamorton replied:
Unfortunately not, and its easy enough to do this:

    with mock_node(stuff) as mn:
        with check_topic(stuff) as ct:
            mn.send(data)
            assert about == ct.message
        with check_topic(stuff) as ct2:
            mn.send(more)
            assert about == ct2.message

irapha replied:
yeah im mainly concerned about when, say, on nodeA receiving a single message, it outputting 2 messages in a row (or some variation of that)

But I'll say that until that is a concern, we won't worry about it.

Just make sure to mention in the readme api that check_topic only receives the first message then stops listening

joshuamorton replied:
mentioned

irapha commented on buzzmobile/tests/test_utils/launch_tools.py:
```diff
@@ -94,9 +149,9 @@ def new_test(self):
                 process.stop()
             if is_master:
-                _launcher = None
+                _LAUNCHER[self.port].stop()
```
I think I was the one who added these `stop`s. Are they needed?

joshuamorton replied:
No clue

irapha replied:
if it works it works

irapha reviewed (changes requested):
I can't make the tests pass locally... It hangs on controller test (not fail, not error, it hangs)

Here's a gist of using `pytest -n1 unit/test --boxed`: https://gist.github.com/7e7e956319bc68cddbb8236dac0b6980

Here's one using `pytest unit/test`: https://gist.github.com/ba4a94ff3eca41a5ab560ab8a9b98552

I was actually able to make tests pass by just running again. Please read follow up comments on the "await" comment below (https://github.com/gtagency/buzzmobile/pull/180/files/541562ae9e5c1529e19e1f03efc7c7e7b2dbfd70#diff-6c736e19c4d374acb71866ba13b75078) vvv

irapha commented:
One last question: should we still run in `--boxed`? Can you update `ci_scripts/unittest`?

And if we're not using `pytest-xdist` can you remove the dep from `setup.py`?

irapha commented on buzzmobile/tests/test_utils/launch_tools.py:
```diff
@@ -44,13 +94,17 @@ def launcher(func):
-            global _LAUNCHER # pylint: disable=global-statement
-            _LAUNCHER = launch
+            if self.port in _LAUNCHER:
+                raise Exception('You are using these incorrectly')
```
Also can you make all these exception messages more descriptive?

joshuamorton replied:
Done, what would be a good exception here?

irapha replied:
'Rosmaster port {} already in use. Make sure with_file_name is the first decorator'

or something like that

irapha committed:
add test info to the readme

irapha committed:
add initial test utils readme

joshuamorton commented:
xdist still lets us do multicprocess testing, if we want that.

joshuamorton committed:
Completing code review.

Small changes: improved error messages, some more documentation, and
other minor things (removed todos and cleaned up code).

joshuamorton committed:
Add a more descriptive error message.

joshuamorton committed:
Start the local readme.

joshuamorton committed:
Style fix.

joshuamorton committed:
Updated sleep to be less time after testing.

The sleep time was three seconds, multiplied out over lots of tests is a
lot, and if roscore is running, the roscore failure will happen almost
immidiately, it only takes significant time if roscore isn't already
functional.

joshuamorton committed:
Finish the rostest utils readme.

joshuamorton committed:
Added timeout functionality.

The listener now has a timeout (default 10 seconds) and should throw an
error if no data is recieved by the MockListener in that time. This will
avoid any kind of awkward spin forever loops that were occuring.

joshuamorton committed:
Actually start the timer

joshuamorton committed:
Improve error message on timeout.

irapha commented on buzzmobile/tests/unit/plan/test_controller.py:
```diff
@@ -1,28 +1,17 @@
-            # fail stating the message has not been sent yet.
-            # Even though there is an await in the bearing test.
-            # Also, it doesn't matter which topic is checked
+        with mock_pub('/joy', Joy, queue_size=None) as joy_node:
```
Can you change this None to 0, since its better?

joshuamorton replied:
done

irapha commented on buzzmobile/tests/test_utils/launch_tools.py:
```diff
@@ -44,13 +100,19 @@ def launcher(func):
+            if self.port in _LAUNCHER:
+                raise RosLaunchException('Rosmaster port {} already in use. '
+                'You must call use @with_launch_file only once for any single '
+                'test.'.format(self.port))
```
Can you make this message also mention that with_launch_file needs to be called before any launch_nodes?

joshuamorton replied:
done

irapha commented on buzzmobile/tests/test_utils/mock_nodes.py:
```diff
@@ -4,114 +4,138 @@
+            timer = threading.Timer(self.timeout, kill_proc)
+            try:
+                timer.start()
+                data = self.proc.stdout.read(sio.len)
```
quick clarification question:

if i start a publisher node, start a listener node, publish a message, and do listener.message *after* that message was sent, will this still capture that? I assume so bc stdin will wait until its consumed, but I just want to make sure that the following would still work:

```python
def test(self):
    with mock_pub('/some') as p:
        with check_topic('/some') as ct:
            p.publish()
            time.sleep(1000)
            assert ct.message
```

joshuamorton replied:
ye, it will.

irapha commented on buzzmobile/tests/test_utils/readme.md:
```diff
@@ -0,0 +1,258 @@
+
+## Quick Notes
+
+ - If your tests are inexplicably failing, run `rm -r *.pyc`, especially if you
```
"... failing *or hanging* ..."

joshuamorton replied:
done

irapha commented on buzzmobile/tests/test_utils/readme.md:
```diff
@@ -0,0 +1,258 @@
+
+ROS requires launch files. Normally you would run something like `roslaunch
+xyz.launch`, but you can't do that when you aren't using the roslaunch command
+line tool, so this will do that for you. 
```
can you include code examples for each of these? Like a single example per thing is fine, just to demonstrate

irapha replied:
actually also maybe detail the api as whatever optianal kwargs params we may have, such as timeouts.

irapha replied:
actually I just saw you have an example below, so feel free to ignore this comment. But can you include all the details in the example (such as timeout optional arg)?

joshuamorton replied:
done.

irapha commented on buzzmobile/tests/test_utils/readme.md:
```diff
@@ -0,0 +1,258 @@
+
+If you're just a user, no need for this section. If, however, you have the
+misfortune of having to modify `rostest_utils`, I offer you my sympathy and my
+knowledge.
```
+1

joshuamorton committed:
Final code review changes.

This should be everything!

irapha committed:
nits

irapha reviewed (approved):
quality readme. merge when tests are passing

irapha commented:
Correct this comment if I'm wrong, but this:

Closes #177 entirely
